File: backend/src/models/Category.ts
import mongoose, { Schema, Document } from 'mongoose';
import { ICategory } from '@mufar-commerce/shared';

export interface ICategoryDocument extends Omit<ICategory, '_id' | 'createdAt' | 'updatedAt'>, Document {}

const categorySchema = new Schema<ICategoryDocument>(
  {
    name: {
      type: String,
      required: [true, 'Category name is required'],
      trim: true,
    },
    slug: {
      type: String,
      unique: true,
      lowercase: true,
      trim: true,
    },
    description: { type: String, trim: true },
    parent: {
      type: Schema.Types.ObjectId,
      ref: 'Category',
      default: null,
    },
    image: { type: String },
    isActive: { type: Boolean, default: true },
    sortOrder: { type: Number, default: 0 },
  },
  {
    timestamps: true,
    toJSON: {
      transform(_doc: any, ret: any) {
        delete ret.__v;
        return ret;
      },
    },
  }
);

categorySchema.pre<ICategoryDocument>('save', function (next) {
  if (this.isModified('name') || !this.slug) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/(^-|-$)/g, '');
  }
  next();
});

categorySchema.index({ slug: 1 });
categorySchema.index({ parent: 1 });

const Category = mongoose.model<ICategoryDocument>('Category', categorySchema);
export default Category;

File: mobile-app/lib/models/category.dart
class Category {
  final String id;
  final String name;
  final String? slug;
  final String? description;
  final String? parentId;
  final String? image;
  final bool isActive;
  final int sortOrder;
  final int? productCount;

  Category({
    required this.id,
    required this.name,
    this.slug,
    this.description,
    this.parentId,
    this.image,
    this.isActive = true,
    this.sortOrder = 0,
    this.productCount,
  });

  factory Category.fromJson(Map<String, dynamic> json) {
    final parent = json['parent'];
    String? parentId;
    if (parent is Map<String, dynamic>) {
      parentId = parent['_id'] as String?;
    } else if (parent is String) {
      parentId = parent;
    }

    return Category(
      id: json['_id'] as String? ?? '',
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String?,
      description: json['description'] as String?,
      parentId: parentId,
      image: json['image'] as String?,
      isActive: json['isActive'] as bool? ?? true,
      sortOrder: json['sortOrder'] as int? ?? 0,
      productCount: json['productCount'] as int?,
    );
  }
}
